### PR TITLE
Introduce ColorMode, StencilMode, DepthMode abstractions

### PR DESCRIPTION
**Pull request by lbud**
Abstracts all depth-, color- and stencil-related state manipulation into "modes," so that each render function is responsible for setting its own modes (with `context.set*Mode` as an intermediate step, but eventually as part of `program.draw`). This eliminates all instances of render functions needing to reset assumed state after they're finished 🎉 

I figured out all of the quirks I mentioned offline, so `context.setStencilMode`, `context.setColorMode`, and `context.setDepthMode` behave almost exactly the same as in native (the one slight difference being how we handle `blendColor`, which is set mostly as default-initialized in native, but is optional and often therefore not set here).

Benchmarks: https://bl.ocks.org/anonymous/raw/5a607b699c5f57ed705e4a7e2aa342a1/

The `LayerBackground` test does consistently look significantly slower (~1.89 ms, as opposed to previously ~0.59 ms) but I don't feel that's concerning given that most maps use one or very few background layer.

I don't feel comfortable merging this yet until I add a bunch of render tests that test various combinations of layers (which would have caught e.g. https://github.com/mapbox/mapbox-gl-js/commit/5565087ddabcc792b5b7ed5a2d500849112bd6be — I'm going to add the render test for that to a new branch off of current master).

Refs https://github.com/mapbox/mapbox-gl-js/issues/145.
Refs https://github.com/mapbox/mapbox-gl-native/pull/10655.

### CODE REVIEW

File: src/render/draw_background.js [hunk 1]
```diff
@@ -27,9 +28,9 @@ function drawBackground(painter: Painter, sourceCache: SourceCache, layer: Backg
     const pass = (!image && color.a === 1 && opacity === 1) ? 'opaque' : 'translucent';
     if (painter.renderPass !== pass) return;
 
-    context.stencilTest.set(false);
-
-    painter.setDepthSublayer(0);
+    context.setStencilMode(StencilMode.disabled());
+    context.setDepthMode(painter.depthModeForSublayer(0, true));
```
[the remark this answers is not in the document]
Comment (lbud) [reply src/render/draw_background.js]: Oh — actually both are wrong. Since we draw opaque backgrounds in the opaque pass we should be writing to the depth mask only on that pass, as `drawFill` does. (I happened to try a test case with two opaque background layers, and they're rendered in the wrong order — only the bottom layer appears. Obviously this is an edge case that probably very few people would ever need to support, but 🤷‍♀️)

File: src/render/painter.js [hunk 2]
```diff
@@ -255,17 +228,31 @@ class Painter {
...
+            return new ColorMode([gl.CONSTANT_COLOR, gl.ONE], new Color(a, a, a, 0), [true, true, true, true]);
+        } else if (this.renderPass === 'opaque') {
+            return ColorMode.unblended();
+        } else {
+            return ColorMode.alphaBlended();
+        }
+    }
+
+    depthModeForSublayer(n: number, mask: DepthMaskType, func: ?DepthFuncType): DepthMode {
+        const farDepth = 1 - ((1 + this.currentLayer) * this.numSublayers + n) * this.depthEpsilon;
```
[the remark this answers is not in the document]
Comment (lbud) [reply src/render/painter.js]: 🤔 I tried doing this and must have been getting something wrong as most things no longer rendered correctly — I'm going to merge this as-is and we can take on https://github.com/mapbox/mapbox-gl-native/issues/9713 separately.
